### ISSUE TITLE
🚀 Feature: Plugin API

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#1457]
We need plugin API(s). 

More to follow.

**Comment by glenjamin** [comment]
Some thoughts on the above proposal:

* While possibly cool, auto-detection of plugins isn't very node-like. Using `--plugin <X>` and requiring the user to list them in `mocha.opts` will be far easier to implement and work basically the same way.
* Perhaps combine with a `--disable-plugin <X>` to allow overriding defaults on the CLI
* I think the module signature should just be a single export function, which receives the mocha plugin API as an argument. This is pretty straightforward and provides a simple way to expose additional functionality.
```js
// my-cool-plugin.js
module.exports = function(mocha) {
  mocha.registerReporter();
  mocha.addCliOption();
  // etc
}
```

I would suggest that core-module-replacement and inter-plugin-communication not be included in an initial release, as they're quite tricky and it would be more important to get the basic mechanisms in first.

A possible initial featureset could be:
* Custom reporters
* Custom interfaces
* Access to the current instance options
* Access to the `Runner` instance (can attach events to do stuff)
* Add extra functions to existing interfaces?

One way to allow plugins to specify extra options but also be parsed from options is to do one pass over argv to get the plugin list, then load them, then do another pass to get the full options list.

**Comment by boneskull** [comment]
@glenjamin Thanks for the input.  I'm going to fiddle with this a bit.

**Comment by jamestalmage** [comment]
First, 
This is a **great** idea.

I've always like chai's plugin architecture:
```javascript
chai.use(require('some-chai-plugin'));
```
That should not preclude functions like `mocha.registerReporter()`, etc. But rather those functions should be called by the plugin author, not the end user. This allows plugin authors to hide ugly implementation details from the user (i.e. my `super-awesome-reporter` may need to install a global `beforeEach`, `afterEach`, etc. You don't want to know that. You just want to to use`super-awesome-reporter`).

So spec should be something like this:
 1. user calls `mocha.use(require('plugin-module'))`

     or optionally `mocha.use(require('plugin-module')(opts))`. 
 2. convention dictates that a plugin module returns a callback.
 3. mocha will validate that callback has never been called before. If it has, silently ignore (mocha should ensure plugins are not added twice - should not be an error).
 4. mocha will call that callback, passing a `registration object` as the first parameter.
 5. registration object will expose all the allowed "hooks"

**Comment by boneskull** [comment]
@jamestalmage `use` sounds ok, but Chai isn't a command-line executable, so that's pretty much what it needs to do.

Mocha, on the other hand, allows CLI options or a `mocha.opts` file, so simply putting `--plugin mocha-foo-reporter` in there might be fine.

What do you think?

Personally, I think `mocha.opts` is kind of lame and would prefer a proper configuration file in JSON and/or YAML format.

**Comment by jamestalmage** [comment]
I wasn't suggesting that you NOT provide a command line. Just suggesting what I thought was a good API. `--plugin mocha-foo-reporter` from the command line, could just become `mocha.use(require('mocha-foo-reporter'));` internally.

I'm not so sure about JSON/YAML. They end up being a bit too restrictive. I like how karma handles config:

```javascript
// karma.conf.js
module.exports = function(karma) {
  karma.set(
    /* ... */
  );
} 
```

That tiny bit of extra verbosity provides all kinds of power:
```javascript
karma.set({
    browsers: [process.env.TRAVIS ? 'Firefox' : 'Chrome']
});
```

This would be super handy for say, skipping certain tests that were part of a PR because they required secure variables on travis (which are not available to PR's for security reasons). Angulars karma configs make [heavy use of this power](https://github.com/angular/angular.js/blob/master/karma-jqlite.conf.js) and provide a good example of an ~~overly~~ complex build.  The 90% of configs that do not need the extra power offered by this approach just end up looking a lot like JSON inside a `module.exports = function` wrapper.

**Comment by glenjamin** [comment]
Fwiw I like that mocha.opts forces all config to also be available on the command line.

**Comment by jamestalmage** [comment]
@glenjamin I don't think there's anything wrong with making all of core config available on the command line, but I don't think you want to force that on every plugin. Certainly provide a mechanism for plugins to digest command line args.

**Comment by boneskull** [comment]
@jamestalmage I'm in disagreement--I understand that as a `.js` file, it gives you more flexibility.  However, my preference is *configuration* instead of, well, *more code*.  I don't want to ever feel compelled to write unit tests against my config files or build scripts. :stuck_out_tongue_closed_eyes: 

> I wasn't suggesting that you NOT provide a command line.

Not sure where I gave you this impression..?

Anyway, I'm of the opinion this sort of thing (and likely other collaborators as well):

> ```js
karma.set({
  browsers: [process.env.TRAVIS ? 'Firefox' : 'Chrome']
});
```

...should not be handled by Mocha; it wades into "task runner"/`Makefile`/shell-script territory.  Detecting an environment variable and changing its behavior is not something Mocha, nor its configuration, should be in the business of.  Instead, a script should run Mocha with different options.  This keeps Mocha simple.  You'll find a similar argument in many other GH issues.

I'd probably support `.json` and `.yaml` config files out-of-the-box.  This would add only a single 3p dep.  That said, as far as supporting a `.js` config: if there is some sort of popular need, I'd be behind adding it.  I certainly have ideas about what should and should not be in Mocha, but that doesn't mean I'm deaf to public opinion.

`mocha.opts` just seems pretty silly to me, as its functionality is duplicated--and improved upon--by a simple shell script.

@glenjamin 

The config file should allow plugin-specific configuration which may or may not be applicable to the command line; think browser-only reporters.

**Comment by jamestalmage** [comment]
@boneskull Yeah, I pretty much anticipated this would be your response, but I'm a masochist, so I'm going to keep arguing :stuck_out_tongue_winking_eye: .

>  I don't want to ever feel compelled to write unit tests against my config files or build scripts.

If you were ever silly enough to make your build script complicated enough where you did feel so compelled, would you rather write unit tests against a bash script, a Makefile, or a javascript node module?

> ...should not be handled by Mocha; it wades into "task runner"/Makefile/shell-script territory

To be clear, in my example all karma sees is `browsers:['Firefox']` or `browsers:['Chrome']`, it never sees an complex logic. As I said:

>  The 90% of configs that do not need the extra power offered by this approach just end up looking a lot like JSON inside a `module.exports = function` wrapper.

As for keeping mocha simple, I couldn't agree more, but I would think a `.js` config is pretty simple.

```javascript
var config = require('config.json');
// vs
var config = require('config.js')(mocha);
```

**Comment by jamestalmage** [comment]
> No, browser reporters won't have access to the config file, so yeah, I don't envision anything in the config file that wouldn't be available on the command line.

Why would you prevent browser reporters from accessing the config file.

**Comment by boneskull** [comment]
> Why would you prevent browser reporters from accessing the config file.

@jamestalmage Because browser reporters run in the browser and do not have access to the filesystem.

**Comment by jamestalmage** [comment]
Yeah, but couldn't you bundle it right alongside all the sources? That's certainly an argument for sticking with JSON.

**Comment by jamestalmage** [comment]
err, nevermind. I'm conflating mochas abilities with `karma-mocha` (which handles bundling the files up for you)

**Comment by pluma** [comment]
It'd be neat if it were possible to use mocha programmatically without it requiring all the tty/shell related dependencies.

I integrated Mocha in ArangoDB to allow us to execute tests in ArangoDB's JavaScript environment (which is mostly compatible with node modules) but had to shim several dependencies as well as some internals like `process.argv` just in order to be able to `require('mocha')`.

I think it'd be a good idea to extract the mocha runner and interfaces and have them separate from the CLI (which could then handle features like plugin auto-loading).

It'd also be nice if the interfaces used `post-require` to clean up after themselves. This would allow re-using the context in which a test suite is run. I had to modify the logic of `mocha.loadFiles` so it would create a new context object (really just a bunch of variables that get dumped in a closure the module code is executed inside of) for the tests to execute in.

See https://github.com/arangodb/arangodb/blob/660e2147da2075efc8e9fa5bf64dc96c77b930c7/js/server/modules/org/arangodb/foxx/mocha.js

**Comment by boneskull** [comment]
@pluma see #1457

**Comment by boneskull** [comment]
oh wait, this *is* #1457.

**Comment by boneskull** [comment]
@pluma Anyway, it's going to be iterative.  I'm not sure we can *release* it iteratively, but the first thing I'm focusing on is providing several common interfaces for reporters of different types.  

Currently I have a `Base` reporter which all other interfaces extend.  Then I have (or will have) three more:

1.  `Console` -- all reporters which work in a TTY should use this interface
2.  `Browser` -- for reporters which work in the browser
3.  `Stream` -- for reporters working with other non-TTY streams

1) may be a specific case of 3), so there's that, but that's kind of what I'm working from.  Then, I propose moving *every reporter except for a default in each category* into its own repository.  That means "spec", "html", and either create a new generic "file" reporter or use "json".

Suggestions welcome!

**Comment by dasilvacontin** [comment]
Feels like the most straight forward place to start out with, and we get to move out the reporters as a bonus. Hopefully I can contribute to this, but my free time is quite limited lately..

**Comment by pluma** [comment]
Shiny. If there's anything I can help with, I'm available. I'd love to see mocha become more environment agnostic.

**Comment by opsb** [comment]
+1 for an API for global hooks, this would be really helpful for setting up/tearing down any datastores/caches.

**Comment by boneskull** [comment]
@opsb You can't do this?

```js
// top of file
beforeEach(function() {
  // blah
});

describe('stuff', function() {
  // more tests
});
```

or perhaps not asynchronously?  I thought this was mentioned in another ticket, but I don't recall where.

**Comment by boneskull** [comment]
anyway

`mocha.opts` (or more precisely, "the configuration file") should be accessible via a programmatic API.

**Comment by boneskull** [comment]
as mentioned in #1668

**Comment by opsb** [comment]
@boneskull I came across that just earlier today, I had no idea!

**Comment by jmeas** [comment]
Just stopping in to say I am a huge fan of the idea of exposing the config file via a programmatic API. Maybe switching it over to JSON (like all of those other rc files out there) would make this easier.

**Comment by perrin4869** [comment]
Maybe [domictarr/rc](https://github.com/dominictarr/rc) could be useful to implementing a configuration file easily

**Comment by boneskull** [comment]
> Maybe domictarr/rc could be useful to implementing a configuration file easily

Definitely.  I forked it to [rc-yaml](https://npmjs.com/package/rc-yaml) which adds, well, YAML support.

**Comment by OmarDarwish** [comment]
Is this issue still being considered?

**Comment by JoshuaKGoldberg** [comment]
I've been linking issues to this one that seem relevant. There are ... quite a few 🫠. But given how many use cases folks have asked for Mocha that would be solveable by a plugin API, they're all required reading to submit a proposal for this.

https://github.com/mochajs/mocha/issues/4116 has some very good notes too.
